README.md
s3qloss
=======

s3ql plugin for aliyun oss

=======

------------------------
Installation:Test
------------------------

>># git clone https://github.com/thefirstwind/s3qloss
>># cd s3qloss


for contribute 
    # vi ~/.s3ql/authinfo
    # chmod 600 ~/.s3ql/authinfo
    * see contrib/authinfo
#>>>>>>>>>
[gs]
storage-url: gs://ossfs
backend-login: GOOGGURDYPMCGGOLVY53
backend-password: VxJASoLVq3N+wJGC6OGWmRdOxg+1EZ4hvUIDCsZU
[oss]
storage-url: oss://ossbucket
backend-login: IJ1a5Kg1ujqsrOfb
backend-password: EKQWar0RJAqrZdeM3jXfeHeWB5LeNf
#>>>>>>>>>


for test
    # vi ~/.s3ql/authinfo2
    # chmod 600 ~/.s3ql/authinfo2
    * see contrib/authinfo2
  
#>>>>>>>>>
[gs-test]
backend-login: GOOGGURDYPMCGGOLVY53
backend-password: VxJASoLVq3N+wJGC6OGWmRdOxg+1EZ4hvUIDCsZU
test-fs: gs://ossfstest
[oss-test]
backend-login: IJ1a5Kg1ujqsrOfb
backend-password: EKQWar0RJAqrZdeM3jXfeHeWB5LeNf
test-fs: oss://ossbucket-test
#>>>>>>>>>


>># python setup.py test


Changes.txt
2013-09-20, s3qloss 1.16.1
  
  * add aliyun oss support by xing xiaoning <thefirstwind@gmail.com>
  
  * change the project name from S3QL to S3QLOSS , means that is S3QL plugin
    for aliyun OSS
     
  
2013-07-12, S3QL 1.16

  * Eliminated a race condition that resulted in an apparent loss of
    remote metadata when the mount.s3ql process was interrupted after
    renaming the remote metadata but before having completely uploaded
    the local metadata.

  * Attempting to run mount.s3ql with --allow-user but without having
    the necessary permissions no longer results in the file system
    being marked as needing fsck.

  * S3QL is now able to detect server errors and redirections before
    all data has been sent to the server. This avoids pointless data
    transfers and should fix a problem with S3QL operating very slow
    when operating on recently created storage buckets/containers.

  * There is a new 'rackspace' backend that allows specification of
    the object storage region in the storage URL. The rackspace
    backend also uses a different authentication API than the swift
    backend (which can still be used for Rackspace as well). When
    using the 'rackspace' backend, use your regular Rackspace password
    instead of the API key as the backend password.

2013-05-27, S3QL 1.15

  * Cycling of metadata backups when using the local backend now
    takes much less time.

  * It is no longer required that S3QL and Python APSW be linked to
    the same SQLite shared library.

  * mount.s3ql: fixed crash when using --metadata-upload-interval 0.

  * Instead of completely unmounting the file system (and leaving the
    mount point inaccessible) when backend data is missing or
    corrupted, S3QL now returns an error and switches to read-only
    operation.

2013-04-09, S3QL 1.14

  * Reduced CPU consumption when S3QL is idle.

  * S3QL now automatically retries requests when S3 reports an
    "internal error" (this functionaly got accidentally broken
    in S3QL 1.13).

  * Fixed a hang when using s3qlrm to delete a directory with
    a very large number of sub-directories.

2013-03-03, S3QL 1.13.2

  * Allow S3 bucket names starting with a number.

  * Return EOPNOTSUP error on ACL operations. The previous value of
    EINVAL works with FUSE, but confuses several applications.

2013-02-24, S3QL 1.13.1

  * Change error code returned to applications trying to set ACL
    extended attributes. The previous error code of ENOSYS had the
    unintended side effect that some FUSE versions gave up trying to
    get/set any extended attributes at all.
  
2013-02-23, S3QL 1.13

  * Fixed issues when trying to access a directory at the root
    of the file system that has recently been removed with s3qlrm.
    (In that case the directory was no longer included in the
    readdir() output, but it was still posssible to stat() it because
    was still present in the kernel's dentry cache).

  * When trying to connect to the backend, S3QL now also retries at
    increasing intervals if DNS appears to be unavailable (i.e., if
    there's currently no network at all). This means that mount.s3ql
    will no longer bail out when called without a network connection
    but simply retry (printing warning messages in the process).

  * Work around undocumented "Timeout" reply from Google Storage and
    retry instead of crashing.

  * Instead of failing with a bogus "Invalid credentials or skewed system
    clock" error when trying to access a bucket with an invalid DNS
    name, S3QL now prints a proper error message.

    (Such bucket names are still allowed when using the US Standard
    regions, but generally not a good idea).

  * Removed ostensible ACL support. It was found that S3QL never kept
    ACLs and file permission bits synchronized as mandated by POSIX.
    This is due to a bug in the FUSE library and cannot easily be
    fixed in S3QL. Consequently, S3QL no longer reports ACL support
    unless a FUSE version with the necessary bugfix version is
    installed.

  * Fixed a bug that caused malformed metadata to be written out when
    using recent eglibc versions (the relevant change in eglibc
    was introduced between eglibc versions 2.13 and 2.17).
    
  
2012-09-03, S3QL 1.12

  * Be more verbose about how to remedy the situation when attempting
    to upgrade file system needing fsck.

  * Fsck now detects and fixes directory entries which do not have an
    entry type stored in their mode field.

2012-05-04, S3QL 1.11.1

  * Fixed crash when using S3 backend.
    (Regression introduced in release 1.11).

  * Increase minimum reported file system size to 1 TB, and work
    around df bug so that size is reported properly.

  * umount.s3ql now produces a more helpful error message if the
    file system has crashed.

2012-04-29, S3QL 1.11

  * S3QL no longer uses the confusing "bucket" term (which is S3
    specific). The "bucket-passphrase" option in the authentication
    file has been renamed to "fs-passphrase".

  * Metadata about stored objects is now split into multiple fields,
    so that no field is longer than 256 bytes. This makes S3QL
    compatible with a wider range of storage providers.

    Thanks to Stuart Wallace <stuart@wapbot.co.uk> for reporting this
    issue.

  * The SWIFT backend now retries if it receives 5xx error codes from
    the remote server, and no longer fails when attempting to upload
    data with an expired auth token. Thanks to Ken
    <forken.liu@delta.com.tw> for the report and initial patch!
  
2012-02-24, S3QL 1.10

  * The s3s, s3cs and gss backends have been removed. Use the new
    --ssl option together with the s3, s3c and gs backends instead.

  * S3QL no longer keeps track of consistency guarantees for the
    different backends. The increasing number of different storage
    providers offering different regions, redundancy and availability
    levels makes this no longer feasible.


tests/t1_backends.py
'''
t1_backends.py - this file is part of S3QL (http://s3ql.googlecode.com)

Copyright (C) 2008-2009 Nikolaus Rath <Nikolaus@rath.org>

This program can be distributed under the terms of the GNU GPLv3.
'''

from __future__ import division, print_function, absolute_import

from s3ql.backends import local, s3, gs, s3c, swift, rackspace, oss
from s3ql.backends.common import (ChecksumError, ObjectNotEncrypted, NoSuchObject,
    BetterBackend)
import ConfigParser
import os
import stat
import tempfile
import time
import unittest2 as unittest

class BackendTestsMixin(object):

    def newname(self):
        self.name_cnt += 1
        # Include special characters
        #return "s3ql_=/_%d" % self.name_cnt
        return "s3ql_t/_%d" % self.name_cnt

    #Test OK
#     def test_write(self):
#           
#          print(">>>>test_write:No1")
#          key = self.newname()
#          print("key: %s" % key)
#           
#          print(">>>>test_write:No2")
#          value = self.newname()
#          print("value: %s" % value)
#          metadata = { 'jimmy': 'jups@42' }
#      
#          print(">>>>test_write:No3")
#          self.assertRaises(NoSuchObject, self.backend.lookup, key)
#          print(">>>>test_write:No4")
#          self.assertRaises(NoSuchObject, self.backend.fetch, key)
#      
#          print(">>>>test_write:No5")
#          with self.backend.open_write(key, metadata) as fh:
#              fh.write(value)
#      
#          time.sleep(self.delay)
#           
#          print(">>>>test_write:No6")
#          with self.backend.open_read(key) as fh:
#              value2 = fh.read()
#           
#          print(">>>>test_write:No7")
#          self.assertEquals(value, value2)
#          print(">>>>test_write:No8")
#          self.assertEquals(metadata, fh.metadata)
#          print(">>>>test_write:No9")
#          self.assertEquals(self.backend[key], value)
#          print(">>>>test_write:No10")
#          self.assertEquals(self.backend.lookup(key), metadata)
#   
#     # Test OK
#     def test_setitem(self):
#         key = self.newname()
#         value = self.newname()
#         metadata = { 'jimmy': 'jups@42' }
#        
#         self.assertRaises(NoSuchObject, self.backend.lookup, key)
#         self.assertRaises(NoSuchObject, self.backend.__getitem__, key)
#        
#         with self.backend.open_write(key, metadata) as fh:
#             fh.write(self.newname())
#         time.sleep(self.delay)
#         self.backend[key] = value
#         time.sleep(self.delay)
#        
#         with self.backend.open_read(key) as fh:
#             value2 = fh.read()
#        
#         self.assertEquals(value, value2)
#         self.assertEquals(fh.metadata, dict())
#         self.assertEquals(self.backend.lookup(key), dict())
#     
#     # Test NG
#     def test_contains(self):
#         key = self.newname()
#         value = self.newname()
#       
#         self.assertFalse(key in self.backend)
#         self.backend[key] = value
#         time.sleep(self.delay)
#         self.assertTrue(key in self.backend)
#       
#     # Test OK
#     def test_delete(self):
#         key = self.newname()
#         value = self.newname()
#         self.backend[key] = value
#         time.sleep(self.delay)
#       
#         self.assertTrue(key in self.backend)
#         del self.backend[key]
#         time.sleep(self.delay)
#         self.assertFalse(key in self.backend)
#       
#      # Test NG
#     def test_clear(self):
#         key1 = self.newname()
#         key2 = self.newname()
#         self.backend[key1] = self.newname()
#         self.backend[key2] = self.newname()
#       
#         time.sleep(self.delay)
# # KEI TEST
#         self.assertEquals(len(list(self.backend)), 2)
#         self.backend.clear()
#         time.sleep(5*self.delay)
#         self.assertTrue(key1 not in self.backend)
#         self.assertTrue(key2 not in self.backend)
#         self.assertEquals(len(list(self.backend)), 0)
 
    # Test NG
    def test_list(self):
       
        keys = [ self.newname() for dummy in range(12) ]
        values = [ self.newname() for dummy in range(12) ]
        for i in range(12):
            self.backend[keys[i]] = values[i]
       
        time.sleep(self.delay)
        self.assertEquals(sorted(self.backend.list()), sorted(keys))
     
    # Test NG
    def test_copy(self):
       
        key1 = self.newname()
        key2 = self.newname()
        value = self.newname()
        
        print(">>>>> self.backend.lookup : %s" % self.backend.lookup)
        print(">>>>test_copy No1 .key1: %s" % key1)
        self.assertRaises(NoSuchObject, self.backend.lookup, key1)
        
        print(">>>>test_copy No2 .key2: %s" % key2)
        self.assertRaises(NoSuchObject, self.backend.lookup, key2)
       
        print(">>>>test_copy No3 .value: %s" % value)
        self.backend.store(key1, value)
        time.sleep(self.delay)
        self.backend.copy(key1, key2)
       
        time.sleep(self.delay)
        self.assertEquals(self.backend[key2], value)
#     
#     # Test NG
#     def test_rename(self):
#      
#         key1 = self.newname()
#         key2 = self.newname()
#         value = self.newname()
#         self.assertRaises(NoSuchObject, self.backend.lookup, key1)
#         self.assertRaises(NoSuchObject, self.backend.lookup, key2)
#      
#         self.backend.store(key1, value)
#         time.sleep(self.delay)
#         self.backend.rename(key1, key2)
#      
#         time.sleep(self.delay)
#         self.assertEquals(self.backend[key2], value)
#         self.assertRaises(NoSuchObject, self.backend.lookup, key1)

# This test just takes too long (because we have to wait really long so that we don't
# get false errors due to propagation delays)
#@unittest.skip('takes too long')
class S3Tests(BackendTestsMixin, unittest.TestCase):
    def setUp(self):
        self.name_cnt = 0
        # This is the time in which we expect S3 changes to propagate. It may
        # be much longer for larger objects, but for tests this is usually enough.
        self.delay = 15

        self.backend = s3.Backend(*self.get_credentials('s3-test'), use_ssl=False)

    def tearDown(self):
        self.backend.clear()

    def get_credentials(self, name):

        authfile = os.path.expanduser('~/.s3ql/authinfo2')
        if not os.path.exists(authfile):
            self.skipTest('No authentication file found.')

        mode = os.stat(authfile).st_mode
        if mode & (stat.S_IRGRP | stat.S_IROTH):
            self.skipTest("Authentication file has insecure permissions")

        config = ConfigParser.SafeConfigParser()
        config.read(authfile)

        try:
            fs_name = config.get(name, 'test-fs')
            backend_login = config.get(name, 'backend-login')
            backend_password = config.get(name, 'backend-password')
        except (ConfigParser.NoOptionError, ConfigParser.NoSectionError):
            self.skipTest("Authentication file does not have test section")

        return (fs_name, backend_login, backend_password)

class S3SSLTests(S3Tests):
    def setUp(self):
        self.name_cnt = 0
        self.delay = 15
        self.backend = s3.Backend(*self.get_credentials('s3-test'), use_ssl=True)
        
class SwiftTests(S3Tests):
    def setUp(self):
        self.name_cnt = 0
        self.delay = 0
        self.backend = swift.Backend(*self.get_credentials('swift-test'))

class RackspaceTests(S3Tests):
    def setUp(self):
        self.name_cnt = 0
        self.delay = 0
        self.backend = rackspace.Backend(*self.get_credentials('rackspace-test'))

class GSTests(S3Tests):
    def setUp(self):
        self.name_cnt = 0
        self.delay = 15
        self.backend = gs.Backend(*self.get_credentials('gs-test'), use_ssl=False)

class OSSTests(S3Tests):
    def setUp(self):
        self.name_cnt = 0
        self.delay = 15
        self.backend = oss.Backend(*self.get_credentials('oss-test'), use_ssl=False)
        
class S3CTests(S3Tests):
    def setUp(self):
        self.name_cnt = 0
        self.delay = 0
        self.backend = s3c.Backend(*self.get_credentials('s3c-test'), use_ssl=False)   

class URLTests(unittest.TestCase):
    
    # access to protected members
    #pylint: disable=W0212
     
    def test_s3(self):
        self.assertEquals(s3.Backend._parse_storage_url('s3://name', use_ssl=False)[2:],
                          ('name', ''))
        self.assertEquals(s3.Backend._parse_storage_url('s3://name/', use_ssl=False)[2:],
                          ('name', ''))
        self.assertEquals(s3.Backend._parse_storage_url('s3://name/pref/', use_ssl=False)[2:],
                          ('name', 'pref/'))
        self.assertEquals(s3.Backend._parse_storage_url('s3://name//pref/', use_ssl=False)[2:],
                          ('name', '/pref/'))

    def test_gs(self):
        self.assertEquals(gs.Backend._parse_storage_url('gs://name', use_ssl=False)[2:],
                          ('name', ''))
        self.assertEquals(gs.Backend._parse_storage_url('gs://name/', use_ssl=False)[2:],
                          ('name', ''))
        self.assertEquals(gs.Backend._parse_storage_url('gs://name/pref/', use_ssl=False)[2:],
                          ('name', 'pref/'))
        self.assertEquals(gs.Backend._parse_storage_url('gs://name//pref/', use_ssl=False)[2:],
                          ('name', '/pref/'))
    def test_oss(self):
        self.assertEquals(oss.Backend._parse_storage_url('oss://name', use_ssl=False)[2:],
                          ('name', ''))
        self.assertEquals(oss.Backend._parse_storage_url('oss://name/', use_ssl=False)[2:],
                          ('name', ''))
        self.assertEquals(oss.Backend._parse_storage_url('oss://name/pref/', use_ssl=False)[2:],
                          ('name', 'pref/'))
        self.assertEquals(oss.Backend._parse_storage_url('oss://name//pref/', use_ssl=False)[2:],
                          ('name', '/pref/'))
                            
    def test_s3c(self):
        self.assertEquals(s3c.Backend._parse_storage_url('s3c://host.org/name', use_ssl=False),
                          ('host.org', 80, 'name', ''))
        self.assertEquals(s3c.Backend._parse_storage_url('s3c://host.org:23/name', use_ssl=False),
                          ('host.org', 23, 'name', ''))
        self.assertEquals(s3c.Backend._parse_storage_url('s3c://host.org/name/', use_ssl=False),
                          ('host.org', 80, 'name', ''))
        self.assertEquals(s3c.Backend._parse_storage_url('s3c://host.org/name/pref', use_ssl=False),
                          ('host.org', 80, 'name', 'pref'))
        self.assertEquals(s3c.Backend._parse_storage_url('s3c://host.org:17/name/pref/', use_ssl=False),
                          ('host.org', 17, 'name', 'pref/'))
                                
class LocalTests(BackendTestsMixin, unittest.TestCase):

    def setUp(self):
        self.name_cnt = 0
        self.backend_dir = tempfile.mkdtemp()
        self.backend = local.Backend('local://' + self.backend_dir, None, None)
        self.delay = 0

    def tearDown(self):
        self.backend.clear()
        os.rmdir(self.backend_dir)

class CompressionTests(BackendTestsMixin, unittest.TestCase):

    def setUp(self):
        self.name_cnt = 0
        self.backend_dir = tempfile.mkdtemp()
        self.plain_backend = local.Backend('local://' + self.backend_dir, None, None)
        self.backend = self._wrap_backend()
        self.delay = 0

    def _wrap_backend(self):
        return BetterBackend(None, 'zlib', self.plain_backend)

    def tearDown(self):
        self.backend.clear()
        os.rmdir(self.backend_dir)

class EncryptionTests(CompressionTests):

    def _wrap_backend(self):
        return BetterBackend('schlurz', None, self.plain_backend)

    def test_encryption(self):

        self.plain_backend['plain'] = b'foobar452'
        self.backend.store('encrypted', 'testdata', { 'tag': True })
        time.sleep(self.delay)
        self.assertEquals(self.backend['encrypted'], b'testdata')
        self.assertNotEquals(self.plain_backend['encrypted'], b'testdata')
        self.assertRaises(ObjectNotEncrypted, self.backend.fetch, 'plain')
        self.assertRaises(ObjectNotEncrypted, self.backend.lookup, 'plain')

        self.backend.passphrase = None
        self.assertRaises(ChecksumError, self.backend.fetch, 'encrypted')
        self.assertRaises(ChecksumError, self.backend.lookup, 'encrypted')

        self.backend.passphrase = 'jobzrul'
        self.assertRaises(ChecksumError, self.backend.fetch, 'encrypted')
        self.assertRaises(ChecksumError, self.backend.lookup, 'encrypted')


class EncryptionCompressionTests(EncryptionTests):

    def _wrap_backend(self):
        return BetterBackend('schlurz', 'zlib', self.plain_backend)


# Somehow important according to pyunit documentation
def suite():
    return unittest.makeSuite(LocalTests)


# Allow calling from command line
if __name__ == "__main__":
    unittest.main()


src/s3ql/backends/gs.py
'''
backends/gs.py - this file is part of S3QL (http://s3ql.googlecode.com)

Copyright (C) Nikolaus Rath <Nikolaus@rath.org>

This program can be distributed under the terms of the GNU GPLv3.
'''

from __future__ import division, print_function, absolute_import
from . import s3c
from s3ql.common import QuietError
import logging
import re

# Pylint goes berserk with false positives
#pylint: disable=E1002,E1101,W0201

log = logging.getLogger("backends.gs")

class Backend(s3c.Backend):
    """A backend to store data in Google Storage
    
    This class uses standard HTTP connections to connect to GS.
    
    The backend guarantees immediate get consistency and eventual list
    consistency.
    """

    use_expect_100c = False
    
    def __init__(self, storage_url, gs_key, gs_secret, use_ssl):
        super(Backend, self).__init__(storage_url, gs_key, gs_secret, use_ssl)

        #self.namespace = 'http://doc.s3.amazonaws.com/2006-03-01'
        self.namespace = 'http://doc.storage.googleapis.com/2010-04-03'

    @staticmethod
    def _parse_storage_url(storage_url, use_ssl):
        hit = re.match(r'^gs://([^/]+)(?:/(.*))?$', storage_url)
        if not hit:
            raise QuietError('Invalid storage URL')

        bucket_name = hit.group(1)
        hostname = '%s.commondatastorage.googleapis.com' % bucket_name
        prefix = hit.group(2) or ''
        port = 443 if use_ssl else 80
        return (hostname, port, bucket_name, prefix)        

    def __str__(self):
        return 'gs://%s/%s' % (self.bucket_name, self.prefix)



src/s3ql/__init__.py
'''
__init__.py - this file is part of S3QL (http://s3ql.googlecode.com)

Copyright (C) 2008-2009 Nikolaus Rath <Nikolaus@rath.org>

This program can be distributed under the terms of the GNU GPLv3.
'''

from __future__ import division, print_function

__all__ = [ 'adm', 'backends', 'block_cache', 'cleanup_manager', 'common',
            'cp', 'ctrl', 'daemonize', 'database', 'deltadump', '_deltadump',
            'fs', 'fsck', 'inode_cache', 'lock', 'mkfs', 'mount', 'ordered_dict',
            'parse_args', 'remove', 'statfs', 'umount', 'VERSION',
            'CURRENT_FS_REV', 'REV_VER_MAP' ]

VERSION = '1.16.1'
CURRENT_FS_REV = 16

# Maps file system revisions to the last S3QL version that
# supported this revision.
REV_VER_MAP = { 15: '1.10',
                14: '1.8.1',
                13: '1.6',
                12: '1.3',
                11: '1.0.1' }
